_posts/2022-09-15-redis-job-queue.adoc

```diff
@@ -0,0 +1,302 @@
+This distinction allows implementing a FIFO queue where we write on the left side and consume from the right side.
+
+To manipulate a Redis list, we need the _group of commands_ associated with this data structure.
+In the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/supes-application/src/main/java/me/escoffier/quarkus/redis/supes/SupesService.java[SupesService class], we inject the `RedisDataSource` and retrieve the group of commands:
+
+[source,java]
+----
+public SupesService(RedisDataSource dataSource, ...) {
+    commands = dataSource.list(FightRequest.class);
+  // ...
+}
+----
+
+Let's now look at the `submitAFight` method:
+
+[source,java]
+----
+public FightRequest submitAFight() {
+    var hero = Hero.getRandomHero();
+    var villain = Villain.getRandomVillain();
+    var id = UUID.randomUUID().toString();
+    var request = new FightRequest(id, hero, villain);
+    commands.lpush("fight-requests", request);
+    return request;
+}
+----
+
+The `submitaFight` method retrieves the random fighters, computes an id,  builds the `FightRequest` instance, and executes the `LPUSH` command.
+The `LPUSH`  command writes the given item to the left side of the list stored at the given key (`fight-requests`).
+
+== Receiving the job requests
+
+Let's now look at the other side: the fight simulators.
+The simulators poll the `FightRequests` from the Redis list modelizing our job queue and simulate the fight.
+
+The simulator is implemented in https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightSimulator.java[`me.escoffier.quarkus.redis.fight.FightSimulator`].
+The constructor receives a configured name (to distinguish multiple simulators) and the Redis data source.
+It creates the objects to emit the Redis commands to read from a Redis list:
+
+[source,java]
+----
+public FightSimulator(@ConfigProperty(name = "simulator-name") String name, Logger logger, RedisDataSource ds) {
+    this.name = name;
+    this.logger = logger;
+    this.queue = ds.list(FightRequest.class);
+    // ...
+}
+----
+
+The simulator polls the fight requests and for each of them simulate the fight.
+The implementation is an _infinite_ loop (it only stops when the application is shut down).
+In each iteration, it reads the pending `FightRequest` from the right side of the queue with the `BRPOP` command.
+If there is no pending request, it restarts from the beginning of the loop.
+If it has a request, it simulates the fight:
+
+[source,java]
+----
+@Override
+public void run() {
+    logger.infof("Simulator %s starting", name);
+    while ((!stopped)) {
+        KeyValue<String, FightRequest> item =
+            queue.brpop(Duration.ofSeconds(1), "fight-requests");
+        if (item != null) {
+            var request = item.value();
+            var result = simulate(request);
+            //...
+        }
+    }
```

Comment:
Not related to this blog post, but maybe we should have some kind of super easy to do this, like an annotation on a method that will be called by Quarkus - something like what Reactive Messaging does.

Reply:
👍🏼